### PR TITLE
Add rust implementation in Statevector.evolve method when operator is Pauli class

### PR DESCRIPTION
**Pull request by YumaNK**
### Summary
I added `_evolve_pauli()` method which is called in `Statevector.evolve()` method when its argument is Pauli class.
The method runs in rust backends for the sake of fast execution.
The rust code is written in the manner of the apply_pauli method in qiskit-aer/src/simulators/statevector/qubitvector.hpp

### Details and comments
Although I placed constant values in the pauli_evolve.rs, one may think that there should be somewhere more appropriate to locate. I plan to make another pull request to manage it and to apply the constant values to a series of codes in qiskit-terra where faster execution is expected.  There seem to be several codes that calculate 2^n explicitly.

## Request for advice
For further improvement, I am trying to parallelize the iterative process in rust.
While I designed `_evolve_pauli` method  as `Statevector.data` evolves in the inplace way for faster execution, rust do not allow parallelize mutable function. 

Currently, the iterative process follows,
```
for i in 0..(n >> 1) {
    lambda(i);
}
```
but if I replace it as `(0..n).into_par_iter().map(lambda);`, it cause compile error as following.
```

    error[E0525]: expected a closure that implements the `Fn` trait, but this closure only implements `FnMut`
       --> src/pauli_evolve.rs:244:26
        |
    244 |         let mut lambda = |i: usize| {
        |                          ^^^^^^^^^^ this closure implements `FnMut`, not `Fn`
    245 |             data_arr[i] *= phase;
        |             -------- closure is `FnMut` because it mutates the variable `data_arr` here
    ...
    251 |         (0..n).into_par_iter().map(lambda);
        |                                --- the requirement to implement `Fn` derives from here

```

I would appreciate it if you could suggest a good workaround.

Mentored by @ikkoham

### CODE REVIEW

**Comment by mtreinish** [comment]
For the compiler error if you're mutating the array while iterating you can use [par_iter_mut](https://docs.rs/rayon/latest/rayon/iter/trait.IntoParallelRefMutIterator.html#tymethod.par_iter_mut) instead of `par_into_iter()`. You might bounce up against the borrow checker doing that though. You probably want to iterate over `data_arr` directly with the mutable parallel iterator to avoid this since it'll track that usage correctly.

File: src/pauli_evolve.rs [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+use std::cmp;
+
+use numpy::PyArray1;
+use numpy::PyReadonlyArray1;
+use pyo3::exceptions::PyOverflowError;
+use pyo3::prelude::*;
+use rayon::prelude::*;
+
+// constant values
+const BITS: [usize; 64] = [
```
Comment (mtreinish) [inline src/pauli_evolve.rs:23]: `usize` can be a 32bit integer when running on 32 bit platforms so the max allowed size would be 4,294,967,295 on those platforms and will lead to a compilation error. There are two ways around this, either keep using `usize` and use `#[cfg(target_pointer_width = "64")]` for this and add a variant for 32 bit with the max value capped to something expressible on 32bits. Or alternatively you can change the type to be `u64` to ensure we can represent all the values but then you'll have to handle conversion to usize (including potential error handling) anywhere you're using this for array access

**Comment by mtreinish** [comment]
The unittest failure looks like it's because the first test job runs with our MSRV of 1.56. Destructuring like `(a, b, c) = foo()` wasn't stabilized until rust 1.59, so you'll have to manually assign things like:

```rust
let res = foo()
a = foo.0
b = foo.1
c = foo.2
```

or a similar pattern to maintain compatibility with rust 1.56. A trick I used when doing local developement is to install the msrv toolchain with rustup `rustup install 1.56` and then test build with `cargo +1.56 build` before pushing just to sanity check (I've tripped over this exact feature not being present in the MSRV in an earlier PR).

**Comment by YumaNK** [comment]
@mtreinish 
Thank you for providing very useful advice on parallelization and the reason for the unittest failure.
This made further code modifications much easier!
I will reflect on your comment in the next commit.

File: src/pauli_evolve.rs (hunk 1 above)
Comment (YumaNK) [reply src/pauli_evolve.rs:23]: If I use `#[cfg(target_pointer_width = "64")]`, do I need to put it on every function that calls the `BITS` or `MASK` variable? Or do I just put it at the beginning line of pauli_evolve.rs?

Comment (mtreinish) [reply src/pauli_evolve.rs:23]: You can just do this at the top (assuming you put an else for 32bit). It's basically just a shorthand for doing an if/else in the compiler scope for the constant definition. An example of this pattern I used in rustworkx you can see here:

https://github.com/Qiskit/rustworkx/blob/c817aa1660b5db5dad64b5d41a5089ca5fa8f22e/src/generators.rs#L975-L979

`MAX_ORDER` there is a globally defined constant and the value depends on target architecture (it's 60 on 64bit targets and 29 on 32bit targets). Then in the `binomial_tree_graph()` function the `MAX_ORDER` variable is just used, the definition of the constant is just different based on the compiler options.

**Comment by HuangJunye** [comment]
Hi @YumaNK, thanks a lot for your contribution. Are you still working on the PR? Please let us know if you need help.

**Comment by ikkoham** [comment]
I will finalize and merge this PR

**Comment by mtreinish** [comment]
This PR is pretty stale at this point so I'm going to close it for now. But please feel free to reopen this if you are going to update this and continue work on it.

**Pull request closed without merging**